tests/cupy_tests/lib_tests/test_polynomial.py

```diff
@@ -423,3 +437,96 @@ def test_polyroutine_diff_types_poly1d(self, xp, dtype1, dtype2):
             out = func(a, b)
         assert out.variable == 'x'
         return out
+
+
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'input': [[2, -1, -2], [-4, 10, 4]],
+}))
+class TestRootsReal(unittest.TestCase):
+
+    @testing.for_signed_dtypes()
+    @testing.numpy_cupy_allclose(rtol=1e-6)
+    def test_roots(self, xp, dtype):
+        a = xp.array(self.input, dtype)
+        out = xp.poly1d(a).roots
+        return xp.sort(out)
+
+
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'input': [[3j, 1.5j, -3j], [3 + 2j, 5], [3j, 0], [0, 3j]],
+}))
+class TestRootsComplex(unittest.TestCase):
+
+    @testing.for_complex_dtypes()
+    @testing.numpy_cupy_allclose(rtol=1e-6)
+    def test_roots(self, xp, dtype):
+        a = xp.array(self.input, dtype)
+        out = xp.poly1d(a).roots
+        return xp.sort(out)
+
+
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'input': [[5, 10], [5, 0], [0, 5], [0, 0], [5]],
+}))
+class TestRootsSpecialCases(unittest.TestCase):
+
+    @testing.for_all_dtypes(no_float16=True, no_bool=True)
+    @testing.numpy_cupy_array_equal()
+    def test_roots(self, xp, dtype):
+        a = xp.array(self.input, dtype)
+        return xp.roots(a)
+
+
+class TestRoots(unittest.TestCase):
+
+    @testing.for_all_dtypes(no_bool=True)
+    @testing.numpy_cupy_array_equal()
+    def test_roots_zero_sized(self, xp, dtype):
+        a = xp.zeros((0,), dtype)
+        return xp.roots(a)
+
+    @testing.for_all_dtypes(no_bool=True)
+    def test_roots_zero_dim(self, dtype):
+        for xp in (numpy, cupy):
+            a = testing.shaped_random((), xp, dtype)
+            with pytest.raises(TypeError):
+                xp.roots(a)
+
+    @testing.for_all_dtypes(no_bool=True)
+    def test_roots_ndim(self, dtype):
+        for xp in (numpy, cupy):
+            a = testing.shaped_arange((3, 1), xp, dtype)
+            with pytest.raises(ValueError):
+                xp.roots(a)
+
+    @testing.for_all_dtypes(no_bool=True)
+    def test_roots_general(self, dtype):
+        a = testing.shaped_arange((5,), cupy, dtype)
+        with pytest.raises(NotImplementedError):
+            cupy.roots(a)
+
+    @testing.for_all_dtypes(no_bool=True)
+    @testing.numpy_cupy_array_equal()
+    def test_roots_zeros1(self, xp, dtype):
+        a = xp.zeros((3,), dtype)
+        return xp.roots(a)
+
+    @testing.for_all_dtypes(no_bool=True)
+    def test_roots_zeros2(self, dtype):
```

Comment:
Rename the test clearer.
```suggestion
    def test_roots_zeros_ndim(self, dtype):
```